lib/sdam/topology.js

```diff
@@ -786,6 +787,8 @@ function parseStringSeedlist(seedlist) {
 
 function topologyTypeFromSeedlist(seedlist, options) {
   const replicaSet = options.replicaSet || options.setName || options.rs_name;
+  if (options.directConnection === false && !replicaSet) return TopologyType.Unknown;
```

Comment:
The original goal of this feature is that drivers (eventually) do auto-discovery of replica sets, unless `directConnection` is specified. I think we want to rework this logic such that we just need to change the default value of `directConnection` in 4.0. Specifically, this line:
```
if (seedlist.length === 1 && !replicaSet) return TopologyType.Single;
```
is problematic.

Perhaps something like this (if I'm reading [this table](https://github.com/mongodb/specifications/blob/master/source/server-discovery-and-monitoring/server-discovery-and-monitoring.rst#initial-topologytype) correctly):
```
if (options.directConnection) {
  return TopologyType.Single;
}

if (replicaSet == null) {
  return TopologyType.Unknown;
}

return TopologyType.ReplicaSetNoPrimary;
```

Reply:
Following tests seem to completely hang and timeout if I seemingly change this function in any way.

* `./test/functional/uri_options_spec.test.js`
* `./test/functional/uri.test.js`

Is it possible the tests need to be edited?

>  1) "before all" hook:
     Error: Server selection timed out after 30000 ms
      at Timeout.waitQueueMember.timer.setTimeout (lib/sdam/topology.js:420:30)
      at Timeout.wrapped [as _onTimeout] (node_modules/wtfnode/index.js:197:27)

Reply:
It's possible, but to clarify the goal of this work:

This project is intended to change the default behaviors for drivers to auto-discover replicasets _by default_. We can't make such a behavioral change in the 3.x series of the driver, since it would be unexpected for our users. 

I think what we need to do is use the `directConnection` uri option as a tool to emulate our legacy behavior. The code I provide above should be pretty close to the ideal behavior at the `Topology` level, so you need to figure out when to inject `directConnection: true` higher in the stack to maintain the legacy behavior. I propose changes need to be made to the connection string parser, so if a seedlist of one is provided we _inject_ `directConnection: true` which will maintain the existing behavior and make our logic in `Topology` more reasonable.